editoast/src/views/timetable/similar_trains/trains_traffic.rs

```diff
@@ -0,0 +1,307 @@
+use arcstr::ArcStr;
+use chrono::DateTime;
+use chrono::Months;
+use chrono::NaiveDate;
+use chrono::Utc;
+use itertools::Itertools;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::fs::File;
+use std::io::BufRead;
+use std::io::BufReader;
+use std::path::PathBuf;
+
+use crate::views::timetable::similar_trains::Waypoint;
+use crate::views::timetable::similar_trains::squash_successive_waypoints;
+
+#[derive(Debug, Deserialize, Clone, Hash, PartialEq, Eq)]
+pub struct JsonTrainOperationalPoint {
+    id: String,
+    stop: Option<bool>,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct JsonTrainTraffic {
+    train_name: String,
+    date: NaiveDate,
+    operational_points: Vec<JsonTrainOperationalPoint>,
+    rolling_stock_name: String,
+    speed_limit_tag: String,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct TrainTraffic {
+    pub id: usize,
+    pub train_name: String,
+    pub start_time: DateTime<Utc>,
+    pub operational_points: Vec<Waypoint>,
+    pub rolling_stock: String,
+    pub speed_limit_tag: String,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+struct Segment {
+    from: Waypoint,
+    to: Waypoint,
+}
+
+#[derive(Clone, Debug)]
+pub struct TrainsTraffic {
+    // Date from which traffic are valid (ie. can be imported)
+    traffic_valid_from: DateTime<Utc>,
+    // Trains by speedlimit
+    trains_by_speedlimit: HashMap<String, HashSet<usize>>,
+    // Trains by rollingstock
+    trains_by_rollingstock: HashMap<String, HashSet<usize>>,
+    // Trains by segment
+    trains_by_segment: HashMap<Segment, HashSet<usize>>,
+    // Train by ID
+    trains_by_id: HashMap<usize, TrainTraffic>,
+}
+
+impl TrainsTraffic {
+    pub fn new(traffic_valid_from: Option<DateTime<Utc>>) -> Self {
+        let traffic_date = if let Some(date) = traffic_valid_from {
+            date
+        } else {
+            Utc::now()
+                .checked_sub_months(Months::new(6))
+                .expect("6 month ago date should be valid")
+        };
+        TrainsTraffic {
+            // Date from which traffic are valid (ie. can be imported)
+            traffic_valid_from: traffic_date,
+            // Trains by speedlimit
+            trains_by_speedlimit: HashMap::new(),
+            // Trains by rollingstock
+            trains_by_rollingstock: HashMap::new(),
+            // Trains by segment
+            trains_by_segment: HashMap::new(),
+            // Train by ID
+            trains_by_id: HashMap::new(),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.trains_by_id.len()
+    }
+
+    /// Load the traffic file
+    #[tracing::instrument(skip_all, level = "info", err, name = "Trains traffic loading")]
```

Comment:
Related to my other comment (at the place where this function is used).


```suggestion
    #[tracing::instrument(skip(self), err, name = "Trains traffic loading", fields(traffic_imported))]
```

**Note 1:** `level = "info"` is not needed since it's the default.
**Note 2:** skip only `self` since we do want `file_path` to be provided as a metadata in the span. `PathBuf` does not implement `Display` but [from the documentation in `tracing::instrument`](https://docs.rs/tracing/latest/tracing/attr.instrument.html), it should fallback on `Debug`. If that doesn't work, you can still do it manually with `skip_all, fields(file_path = ?file_path)`
**Note 3:** at the end of the function, you provide the value for the field `traffic_imported` that has only been initialized as a field so far, but not been provided a value. You can do that with the following.

```rust
tracing::Span::current().record("traffic_imported", self.len());
```